src/group_theory/perm/fin.lean

```diff
@@ -62,3 +63,158 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
   @finset.univ (perm $ fin n.succ) _ = (finset.univ : finset $ fin n.succ × perm (fin n)).map
   equiv.perm.decompose_fin.symm.to_embedding :=
 (finset.univ_map_equiv_to_embedding _).symm
+
+section cycle
+
+/-! ### `cycle` section
+
+Define the permutations `fin.cycle_all n`, the cycle `(0 1 2 ... (n - 1 : fin n))`
+and `fin.cycle_range i`, the cycle `(0 1 2 ... i)`.
+-/
+
+namespace fin
+
+open equiv.perm
+
+/-- `fin.cycle_all` is the cycle `(0 1 2 ... fin.last n)`.
+
+If `n` is `0` or `1`, this will be the identity permutation,
+otherwise it sends `i` to `i + 1` and `fin.last n` to `0`.
+-/
+def cycle_all : Π (n : ℕ), perm (fin n)
+| 0 := 1
+| (nat.succ n) := let inst := fin.comm_ring n in by exactI equiv.add_right 1
+
+@[simp] lemma cycle_all_zero : cycle_all 0 = 1 := rfl
+
+@[simp] lemma cycle_all_one : cycle_all 1 = 1 :=
+subsingleton.elim _ _
+
+lemma cycle_all_succ_apply {n : ℕ} (i : fin n.succ) :
+  cycle_all n.succ i = i + 1 :=
+rfl
+
+@[simp] lemma cycle_all_apply_zero {n : ℕ} : cycle_all n.succ 0 = 1 :=
+by rw [cycle_all_succ_apply, zero_add]
+
+@[simp] lemma cycle_all_apply_last {n : ℕ} : cycle_all n.succ (last n) = 0 :=
+by { ext, rw [cycle_all_succ_apply, last_add_one] }
+
+lemma coe_cycle_all_of_ne_last {n : ℕ} {i : fin n.succ} (h : i ≠ last n) :
+  (cycle_all n.succ i : ℕ) = i + 1 :=
+begin
+  rw cycle_all_succ_apply,
+  have : (i : ℕ) < n := lt_of_le_of_ne (nat.succ_le_succ_iff.mp i.2) (coe_injective.ne h),
+  exact coe_add_one this
+end
+
+lemma coe_cycle_all {n : ℕ} (i : fin n.succ) :
+  (cycle_all n.succ i : ℕ) = if i = fin.last n then 0 else i + 1 :=
+begin
+  split_ifs with h,
+  { simp [cycle_all_succ_apply, h] },
+  exact coe_cycle_all_of_ne_last h
+end
+
+lemma cycle_all_succ {n : ℕ} :
+  cycle_all n.succ = decompose_fin.symm (1, cycle_all n) :=
+begin
+  ext i,
+  cases n, { simp },
+  refine fin.cases _ (λ i, _) i,
+  { simp },
+  rw [coe_cycle_all, decompose_fin_symm_apply_succ,
+      if_congr (show i.succ = last (n + 1) ↔ i = last n, from _) rfl rfl],
```

Comment:
I think this `show` might be a good lemma to extract, `succ_eq_last`

Reply:
It seems a rather situational lemma, but I'm happy to extract it :)

Reply:
It's only worth extracting if you also use the new lemma here!